sql : sequence value out of bounds

mosquito2333 opened the pull request:
Previously, an additional out-of-range sequence value may have been
generated.
The original criterion was whether the previous endSequence value exceeded the limit.
However, when the current endSequence is 1, the increment is 2 and the MaxValue is 2,
the previous endSequence value is 1 and the program does not report an error.
However, when the current endSequenceValue 3 exceeds the limit, an error should be reported

In incrementSequenceUsingCache, the criterion for determining whether a cache's data is eligible is modified to whether the first data that may be stored in the cache exceeds the limit.

Fixes #74127.

Release note (bug fix): Fixed a bug where sequences could return values that are out-of-bounds in some cases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/81123)
<!-- Reviewable:end -->

mosquito2333 commented:
@[otan](https://github.com/otan)  Can you help me review my code?

rafiss reviewed (changes requested):
the code and tests look good! just needs minor changes

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81123#-:-N2HXvl4EEGs5Iav_Bwv:b-bsjll5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mosquito2333 and @rafiss)

___
*[`-- commits` line 12 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81123#-N2HY-13-IjGUD-maV7p):*
please add a line below:

```
Release note (bug fix): Fixed a bug where sequences could return values that are out-of-bounds in some cases.
```
___
*[`pkg/sql/logictest/testdata/logic_test/sequences` line 2218 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/81123#-N2HY9TB2lP53nYnRLj8:-N2HY9TB2lP53nYnRLj9:b-lwj3ua) ([raw file](https://github.com/cockroachdb/cockroach/blob/a83045cae9e33267e0d0f673188b7489088b2660/pkg/sql/logictest/testdata/logic_test/sequences#L2218)):*
> ```
> END
> 
> # nextval allowed a sequence to return values that are out-of-range before
> ```

nit: adjust this comment to:
```
# Regression test for #74127. Assert that out-of-range sequence values are not returned.
```


<!-- Sent from Reviewable.io -->

mosquito2333 committed:
Previously, an additional out-of-range sequence value may have been generated.
The original criterion was whether the previous endSequence value exceeded the limit.
However, when the current endSequence is 1, the increment is 2 and the MaxValue is 2,
the previous endSequence value is 1 and the program does not report an error.
However, when the current endSequenceValue 3 exceeds the limit, an error should be reported

In incrementSequenceUsingCache, the criterion for determining whether a cache's data is eligible is modified to whether the first data that may be stored in the cache exceeds the limit.

Fixes #74127.

Release note (bug fix): Fixed a bug where sequences could return values that are out-of-bounds in some cases.

rafiss reviewed (approved):
the failing tests are flakes or CI problems that are resolved. thanks for your contribution!

bors r+

> Reviewed 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/81123#-:-N2qeQoWBewcUz2aShRy:b-hgq6zk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mosquito2333)



<!-- Sent from Reviewable.io -->
